### PR TITLE
[CARBONDATA-747] Add simple performance test for spark2.1 carbon integration

### PR DESCRIPTION
**Pull request by jackylk**
1. add a CompareTest for spark-2.1 integration: comparing parquet and carbon for scan query and filter query. 
2. add single_pass, dictionary_exlcude, dictionary_include support for dataframe.write to carbon

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1009/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1010/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      "select count(*) from $table",
+      "select sum(c4) from $table",
+      "select sum(c4), sum(c5) from $table",
+      "select sum(c4), sum(c5), sum(c6) from $table",
+      "select sum(c4), sum(c5), sum(c6), sum(c7) from $table",
+      "select sum(c4), sum(c5), sum(c6), sum(c7), avg(c8) from $table",
+      "select * from $table where id = 'i9999999' ",
+      "select * from $table where country = 'p9' ",
+      "select * from $table where city = 'j99' ",
+      "select * from $table where c4 < 1000 "
```
Comment (jarray888) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: please add more testcase, for example:
 "select sum(c4) from $table where id like 'i1%' "
 "select sum(c4) from $table where id like '%10' "
 "select sum(c4) from $table where id like '%xyz%' "

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  private def generateDataFrame(spark: SparkSession): DataFrame = {
+    import spark.implicits._
+    spark.sparkContext.parallelize(1 to 10 * 1000 * 1000, 4)
+        .map(x => ("i" + x, "p" + x % 10, "j" + x % 100, x, x + 1, (x + 7) % 21, (x + 5) / 43, x
+            * 5))
+        .toDF("id", "country", "city", "c4", "c5", "c6", "c7", "c8")
+  }
+
+  private def loadParquetTable(spark: SparkSession, input: DataFrame): Long = timeit {
+    input.write.mode(SaveMode.Overwrite).parquet(parquetTableName)
```
Comment (jarray888) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: suggest to use last char of id column to do partition on parquet, so the comparison is fare.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+// scalastyle:off println
+object CompareTest {
+
+  val parquetTableName = "comparetest_parquet"
+  val carbonTableName = "comparetest_carbon"
+
+  private def generateDataFrame(spark: SparkSession): DataFrame = {
+    import spark.implicits._
+    spark.sparkContext.parallelize(1 to 10 * 1000 * 1000, 4)
+        .map(x => ("i" + x, "p" + x % 10, "j" + x % 100, x, x + 1, (x + 7) % 21, (x + 5) / 43, x
```
Comment (jarray888) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: To simulate a real-life data, please make the data unsorted, like
`map(x => ("i" + randon number, "p" + x % 13, "j" + x % 97, ...)`

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            * 5))
+        .toDF("id", "country", "city", "c4", "c5", "c6", "c7", "c8")
```
Comment (jarray888) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: can you add a column using decimal data type?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1016/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1017/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    CarbonProperties.getInstance()
+        .addProperty("carbon.number.of.page.in.blocklet.column", "10")
+        .addProperty("carbon.enable.vector.reader", "true")
+        .addProperty("enable.unsafe.sort", "true")
+    import org.apache.spark.sql.CarbonSession._
+    val spark = SparkSession
+        .builder()
+        .master("local")
+        .enableHiveSupport()
+        .getOrCreateCarbonSession("./carbonstore")
```
Comment (chenliang613) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: Please put carbonstore to target directory.

Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: ok

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala (hunk 1 above)
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: added

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala (hunk 3 above)
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: ok

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala (hunk 4 above)
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: ok, I found decimal is not supported for dataframe.write, I will  raise a JIRA

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala (hunk 2 above)
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1027/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1037/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1087/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1101/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1117/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+  // | c4          | short     | NA          | measure     | no         |
+  // +-------------+-----------+-------------+-------------+------------+
+  // | c5          | int       | NA          | measure     | no         |
+  // +-------------+-----------+-------------+-------------+------------+
+  // | c6          | big int   | NA          | measure     | no         |
+  // +-------------+-----------+-------------+-------------+------------+
+  // | c7          | double    | NA          | measure     | no         |
+  // +-------------+-----------+-------------+-------------+------------+
+  // | c8          | double    | NA          | measure     | no         |
+  // +-------------+-----------+-------------+-------------+------------+
```
Comment (ravipesala) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: I guess it is better to add some more dimensions with little more cardinality. Now there are only 3 dimensions and one with very high cardinality and others with very less cardinality. So it is better to add some more dimensions with medium cardinality.

**Comment by ravipesala** [comment]
Can you run all the queries one after another. Like run all parquet queries first then do System.gc and wait for few seconds to clear all gc. Then start running carbon queries.
In this we can make sure that one engine memory consumption would not effect other engine.

**Comment by jackylk** [comment]
I am afraid that 10Million record (around 200 MB) is too less, it will fit in OS cache if I run all queries for one after another. Any suggestion on this? Make more data?

**Comment by ravipesala** [comment]
Yes, it fits in OS cache, that is the same case with parquet as well so it would be fare comparison.  What I am suggesting is now we run the queries of parquet and carbon together. So there is a possibility of one engine jvm memory consumption can effect the other engine. It would be better to run all the queries of one engine  and flush the GC and then start running other engine queries.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala (hunk 6 above)
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: ok

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+    println(s"start running queries for $tableName...")
+    var result: Array[Row] = null
+    queries.zipWithIndex.map { case (query, index) =>
+      val sqlText = query.sqlText.replace("$table", tableName)
+      print(s"running query ${index + 1}: $sqlText ")
+      val rt = time {
+        result = spark.sql(sqlText).collect()
+      }
+      println(s"=> $rt sec")
+      (rt, result)
```
Comment (ravipesala) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: Don't cache the result, it is leading to memory issues if more than 2 million data. Just keep the length of the data is enough

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+          s""""timestamp": "$timestamp" """ +
+          s""""date": "${formatter.format(date)}" """ +
+          "}"
+      )
+    }
+  }
+
+  def main(args: Array[String]): Unit = {
+    CarbonProperties.getInstance()
+        .addProperty("carbon.number.of.page.in.blocklet.column", "10")
```
Comment (ravipesala) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: This property is remove now. so it is not required.
Now the size calculated in mb with property `carbon.blockletgroup.size.in.mb` and default size is 64. so no need to configure here.

Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: ok

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala (hunk 7 above)
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CompareTest.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1193/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1194/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1195/

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**